larpandoracontent/LArControlFlow/StitchingCosmicRayMergingTool.cc

```diff
@@ -764,35 +674,58 @@ void StitchingCosmicRayMergingTool::CalculateX0(const PfoToLArTPCMap &pfoToLArTP
 
             // Calculate X0 for the closest pair of vertices
             LArPointingCluster::Vertex pointingVertex1, pointingVertex2;
-
             try
             {
                 LArStitchingHelper::GetClosestVertices(*pLArTPC1, *pLArTPC2, pointingCluster1, pointingCluster2,
                     pointingVertex1, pointingVertex2);
 
-                PfoToPointingVertexMap::iterator pfoToPointingVertexMapIter1(pfoToPointingVertexMap.find(pPfo1));
-                if (pfoToPointingVertexMapIter1 == pfoToPointingVertexMap.end())
+                PfoToPointingVertexMatrix::iterator pfoToPointingVertexMatrixIter1(pfoToPointingVertexMatrix.find(pPfo1));
```

Comment:
Maybe this is complex enough that we need a comment about here saying what the following code block is doing?

Reply:
Yes, I can't get my head around what is actually happening here. This seems like a fairly significant side-effect for a function with the primary(?) goal to compute x0.

Reply:
I think this is the most complex part of the stitching algorithm. In the original version, this code block filled a map (pfo ->  vertex) when a pfo was used in the calculate x0 function. The vertex here is the 'stitching vertex' it is the vertex of the pfo used in the stitching process. This is all 'simple' when you are stitching two pfos because each pfo has one vertex. The map is used later on to access the stitching vertex for each pfo as a means to check which side of the TPC boundary the pfo lies and therefore which direction it should be shifted. But this code block also made sure that no pfo was stitched twice, which is what we want to change. 
So when this is extended each pfo can now have two vertices which the old code cannot deal with. So, now i create a map of maps which i call a matrix. This is a map of (pfo -> (matched pfo -> vertex of the first key pfo)). The code checks that the vertices used to calculate the x0 for each match doesn't change, i.e. a match is not repeated. I thought this was the best extension to the previous checks but thinking about it now i disagree. I think i should change this so that the same vertex is not used twice - which is slightly different in the code. Anyway, this map is really important as it allows us to access the stitching vertex for each pfo later on (in order to determine its shifting direction).

Reply:
Ok, thanks. In that case I wonder if the map could be simplified. Rather than pfo -> matched pfo -> vertex, could you just have pfo -> pair(LArTPC, LArTPC), so that you have a reference that essentially checks if a PFO vertex has been used to stitch across a particular TPC boundary already? As presumably that's actually what we care about here? If that's the case, I think it could make the logic much easier to follow here, and the map key/value pairs simpler.